apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/SimpleManagedTracerTest.cs

```diff
@@ -496,16 +498,178 @@ public void MultipleSpans_Dispose()
+                    span.Dispose();
+                });
+            mockConsumer.Verify();
+        }
+
+        [Fact]
+        public async Task DisjointThreads_Annotate_Fail()
+        {
+            var mockConsumer = new Mock<IConsumer<TraceProto>>();
+            var tracer = SimpleManagedTracer.Create(mockConsumer.Object, ProjectId, TraceId);
+
+            ISpan span = null;
+            await RunInDisjointThreads(() => span = tracer.StartSpan("span-name"), () =>
+            {
+                tracer.SetStackTrace(FilledStackTrace);
+                span.Dispose();
+            }, new InvalidOperationException());
+            mockConsumer.Verify(c => c.Receive(It.IsAny<IEnumerable<TraceProto>>()), Times.Never);
+        }
+
+        [Fact]
+        public async Task DisjointThreads_MultipleSpans()
+        {
+            var mockConsumer = new Mock<IConsumer<TraceProto>>();
+            var tracer = SimpleManagedTracer.Create(mockConsumer.Object, ProjectId, TraceId);
+
+            ISpan span = null;
+
+            Predicate<IEnumerable<TraceProto>> singleChild = t =>
+            {
+                var singleSpan = t.Single().Spans.Single();
+                return IsValidSpan(singleSpan, "child-two", span.SpanId());
+            };
+            Predicate<IEnumerable<TraceProto>> childParent = t =>
+            {
+                var spans = t.Single().Spans;
+                return spans.Count == 2 &&
+                    IsValidSpan(spans[0], "child-one", spans[1].SpanId) &&
+                    IsValidSpan(spans[1], "parent");
+            };
+            mockConsumer.Setup(c => c.Receive(Match.Create(singleChild)));
+            mockConsumer.Setup(c => c.Receive(Match.Create(childParent)));
+
+            await RunInDisjointThreads(() =>
+            {
+                span = tracer.StartSpan("parent");
+                tracer.StartSpan("child-one").Dispose();
+            }, () =>
+            {
+                var tracer2 = span.CreateManagedTracer();
+                tracer2.StartSpan("child-two").Dispose();
+                span.Dispose();
+            });
+
+            mockConsumer.VerifyAll();
+        }
+
+        /// <summary>
+        /// Runs two actions in separate disjoint threads.  The first thread will always finish before
+        /// the second.
+        /// </summary>
+        /// <param name="actionOne">The first action to run.</param>
+        /// <param name="actionTwo">The second action to run.  If expectedException is not null, this action
+        ///     is expected to throw an exception with the type of expectedException.</param>
+        /// <param name="expectedException">Optional, an exception excepted to be thrown in actionTwo.</param>
+        private async Task RunInDisjointThreads(Action actionOne, Action actionTwo, Exception expectedException = null)
+        {
+            var tcs1 = new TaskCompletionSource<bool>();
+            var tcs2 = new TaskCompletionSource<bool>();
```

Comment:
I think you can do away with these and just use `Task.Run` instead of creating threads.